lib/Fhp/Model/TanRequestChallengeFlicker/FlickerTanSvgRenderer.php

```diff
@@ -0,0 +1,150 @@
+            'rx' => 7.5,
+            'ry' => 7.5,
+            'fill' => 'black',
+        ]);
+        // triangle for aiming help for hardware tan generator
+        $doc[] = $this->buildNode('polygon', [
+            'points' => [
+                '25,18', // middle bottom
+                '32,5', // top right
+                '18,5', // top left
+            ],
+            'fill' => 'grey',
+        ]);
+        $doc[] = $this->buildNode('polygon', [
+            'points' => [ // x shifted by 160
+                '185,18', // middle bottom
+                '192,5', // top right
+                '178,5', // top left
+            ],
+            'fill' => 'grey',
+        ]);
+
+        // init flicker rectangles
+        for ($i = 0; $i < 5; ++$i) {
+            $doc[] = $this->buildNode('rect', [
+                'x' => 40 * $i + 10,
+                'y' => 20,
+                'width' => 30,
+                'height' => 75,
+            ], [$this->getAnimation($i)]);
+        }
+        $docAttr = [
+            'xmlns' => 'http://www.w3.org/2000/svg',
+            'width' => $width,
+            'height' => $height,
+            'viewBox' => '0 0 210 105',
+            'preserveAspectRatio' => 'none',
+            'id' => $id,
+        ];
+        $this->svg = $this->buildNode('svg', $docAttr, $doc);
+    }
+
+    /**
+     * @param int $channelNumber flickerRectangles numbered from left to right, 0 is clock, 1 is 2^0, ..., 4 is 2^3 half byte representation
+     */
+    private function getAnimation(int $channelNumber): string
+    {
+        $timePerHalfByte = 1 / ($this->frequency) * 2;
+        $attr = [
+            'attributeName' => 'fill',
+            'calcMode' => 'discrete',
+            'repeatCount' => 'indefinite',
+        ];
+        if ($channelNumber === 0) {
+            // first rectangle is the clock
+            $attr['values'] = 'white;black;white';
+            $attr['keyFrames'] = '0;0.5;1';
+            $attr['dur'] = $timePerHalfByte . 's';
+        } else {
+            // arrange keyframes and colors
+            $colors = array_map(static function (string $pattern) use ($channelNumber) {
+                return $pattern[$channelNumber - 1] === '1' ? 'white' : 'black';
+            }, $this->bitPattern);
+            $keyFrames = range(0, 1, 1.0 / count($this->bitPattern));
+
+            $attr['values'] = implode(';', $colors);
+            $attr['keyFrames'] = implode(';', $keyFrames);
+            $attr['dur'] = ($timePerHalfByte * count($this->bitPattern)) . 's';
+        }
+        return $this->buildNode('animate', $attr);
+    }
+
+    /**
+     * @param string $tag the tag of the node
+     * @param array $attributes name-value pairs of the node attributes
+     * @param string[] $childs of the node
+     * @return string the string representation of the whole node with cilds
+     */
+    private function buildNode(string $tag, array $attributes = [], array $childs = []): string
```

Comment:
There are official PHP extensions like php-simplexml or (more convenient here) php-xmlwriter. Example code: https://www.php.net/manual/en/example.xmlwriter-oop.php

The upside is that this code gets nicer and more secure (proper escaping of all values), the downside is that users need to install/activate that PHP extension if they haven't already.

Reply:
It seems like it needs to delete the first xml tag, but otherwise it seems fine on first sight. Let me know if i should use it.

Reply:
from the injection point of view I don't think it is that big of an issue - if I/we add a sanitzer for the byte array. The rest is static and no user / dev input.

Reply:
i added the input validation for the array